File: research/create_excel.py
# -*- coding: utf-8 -*-
"""01_uranai_list.md + 02_uranai_logic.md + 04_output_type.md を結合し uranai_master.xlsx を生成する。

依存: pip install openpyxl

04 は node research/build_output_type.mjs で再生成可能。
同内容を Node で生成する場合: npm install --prefix research exceljs; node research/create_excel.mjs
"""
from __future__ import annotations

import re
from pathlib import Path

from openpyxl import Workbook
from openpyxl.styles import Alignment, Font, PatternFill
from openpyxl.utils import get_column_letter

ROOT = Path(__file__).resolve().parent
FILE_01 = ROOT / "01_uranai_list.md"
FILE_02 = ROOT / "02_uranai_logic.md"
FILE_04 = ROOT / "04_output_type.md"
OUT_XLSX = ROOT / "uranai_master.xlsx"

HEADERS = [
    "No.",
    "占い名",
    "発祥地・文化圏",
    "カテゴリ",
    "必要な入力情報",
    "判定ロジック概要",
    "出力内容",
    "更新頻度",
    "実装難易度",
    "出力方式",
    "出力方式詳細",
    "実装メモ",
    "概要メモ",
]

HEADER_FILL = PatternFill(start_color="CCCCCC", end_color="CCCCCC", fill_type="solid")
ALT_ROW_FILL = PatternFill(start_color="F5F5F5", end_color="F5F5F5", fill_type="solid")
FONT = Font(name="Meiryo UI", size=11)
FONT_BOLD = Font(name="Meiryo UI", size=11, bold=True)
ALIGN_WRAP_TOP = Alignment(vertical="top", wrap_text=True)
ALIGN_HEADER = Alignment(horizontal="center", vertical="center", wrap_text=True)


def parse_table_rows(body: str) -> dict[str, str]:
    out: dict[str, str] = {}
    for line in body.splitlines():
        line = line.strip()
        if not line.startswith("|"):
            continue
        cells = [c.strip() for c in line.strip("|").split("|")]
        if len(cells) < 2:
            continue
        key, val = cells[0], cells[1]
        if key in ("項目", "------", ""):
            continue
        out[key] = val
    return out


def parse_01(text: str) -> list[dict[str, str]]:
    text = text.replace("\r\n", "\n").replace("\r", "\n")
    chunks = text.split("\n##### ")[1:]
    entries = []
    for ch in chunks:
        first_nl = ch.find("\n")
        if first_nl == -1:
            body = ""
        else:
            body = ch[first_nl + 1 :]
        d = parse_table_rows(body)
        entries.append(d)
    return entries


def parse_02(text: str) -> list[dict[str, str]]:
    text = text.replace("\r\n", "\n").replace("\r", "\n")
    parts = text.split("\n### ")[1:]
    entries = []
    for p in parts:
        first_nl = p.find("\n")
        body = p[first_nl + 1 :] if first_nl != -1 else ""
        d = parse_table_rows(body)
        entries.append(d)
    return entries


def parse_04(text: str) -> dict[str, tuple[str, str]]:
    """| 占い名 | 出力方式 | 詳細 | → { 占い名: (方式, 詳細) }"""
    text = text.replace("\r\n", "\n").replace("\r", "\n")
    out: dict[str, tuple[str, str]] = {}
    for line in text.splitlines():
        line = line.strip()
        if not line.startswith("|"):
            continue
        if "占い名" in line and "出力方式" in line:
            continue
        if re.match(r"^\|\s*[-:|]+\s*\|", line):
            continue
        cells = [c.strip() for c in line.strip("|").split("|")]
        if len(cells) < 2:
            continue
        name, typ = cells[0], cells[1]
        det = cells[2] if len(cells) > 2 else ""
        if not name:
            continue
        out[name] = (typ, det)
    return out


def col_width_units(text: str, max_w: int = 50) -> float:
    """日本語混在でも概算。Excel列幅の上限 max_w。"""
    if not text:
        return 8.0
    w = 0.0
    for ch in str(text):
        if ord(ch) < 128:
            w += 1.0
        else:
            w += 2.0
    return min(max(w * 0.9 + 2, 8), max_w)


def main() -> None:
    t1 = FILE_01.read_text(encoding="utf-8")
    t2 = FILE_02.read_text(encoding="utf-8")
    t4 = FILE_04.read_text(encoding="utf-8")

    list_01 = parse_01(t1)
    list_02 = parse_02(t2)
    out_type = parse_04(t4)

    if len(list_01) != len(list_02):
        raise SystemExit(
            f"件数不一致: 01={len(list_01)} 02={len(list_02)}。同じ並びで揃えてください。"
        )

    wb = Workbook()
    ws = wb.active
    ws.title = "占い一覧"

    for col, h in enumerate(HEADERS, start=1):
        c = ws.cell(row=1, column=col, value=h)
        c.font = FONT_BOLD
        c.fill = HEADER_FILL
        c.alignment = ALIGN_HEADER

    for i, (d1, d2) in enumerate(zip(list_01, list_02), start=1):
        r = i + 1
        name = d2.get("占い名") or d1.get("占い名") or ""
        ot = out_type.get(name, ("", ""))
        if name and name not in out_type:
            print(f"警告: 04_output_type.md に占い名がありません: {name!r}")
        row_vals = [
            i,
            name,
            d1.get("発祥地・文化圏", ""),
            d1.get("出力カテゴリ", ""),
            d2.get("必要な入力情報", ""),
            d2.get("判定ロジック概要", ""),
            d2.get("出力内容", ""),
            d2.get("更新頻度", ""),
            d2.get("実装難易度", ""),
            ot[0],
            ot[1] if ot[0] == "C" else "",
            d2.get("実装メモ", ""),
            d1.get("概要メモ", ""),
        ]

        is_alt = (r % 2 == 0)
        for col, val in enumerate(row_vals, start=1):
            c = ws.cell(row=r, column=col, value=val)
            c.font = FONT
            c.alignment = ALIGN_WRAP_TOP
            if is_alt:
                c.fill = ALT_ROW_FILL

    for col in range(1, len(HEADERS) + 1):
        letter = get_column_letter(col)
        maxlen = 8.0
        for row in range(1, ws.max_row + 1):
            v = ws.cell(row=row, column=col).value
            maxlen = max(maxlen, col_width_units(v if v is not None else ""))
        ws.column_dimensions[letter].width = maxlen

    ws.freeze_panes = "A2"
    wb.save(OUT_XLSX)
    print(f"Wrote {OUT_XLSX} ({ws.max_row - 1} data rows)")


if __name__ == "__main__":
    main()

File: research/create_excel.mjs
/**
 * 01_uranai_list.md + 02_uranai_logic.md + 04_output_type.md → uranai_master.xlsx
 * Run: node research/create_excel.mjs
 * 04: node research/build_output_type.mjs
 */
import fs from "fs";
import { dirname, join } from "path";
import { fileURLToPath } from "url";
import ExcelJS from "exceljs";

const __dirname = dirname(fileURLToPath(import.meta.url));
const BASE = __dirname;
const FILE_01 = join(BASE, "01_uranai_list.md");
const FILE_02 = join(BASE, "02_uranai_logic.md");
const FILE_04 = join(BASE, "04_output_type.md");
const OUT = join(BASE, "uranai_master.xlsx");

function parseTableRows(body) {
  const out = {};
  for (const line of body.split(/\r?\n/)) {
    const t = line.trim();
    if (!t.startsWith("|")) continue;
    const cells = t
      .replace(/^\|/, "")
      .replace(/\|$/, "")
      .split("|")
      .map((c) => c.trim());
    if (cells.length < 2) continue;
    const [key, val] = cells;
    if (key === "項目" || key === "------" || !key) continue;
    out[key] = val;
  }
  return out;
}

function parse01(text) {
  const chunks = text.split(/\r?\n##### /).slice(1);
  return chunks.map((ch) => {
    const nl = ch.indexOf("\n");
    const body = nl === -1 ? "" : ch.slice(nl + 1);
    return parseTableRows(body);
  });
}

function parse02(text) {
  const chunks = text.split(/\r?\n### /).slice(1);
  return chunks.map((ch) => {
    const nl = ch.indexOf("\n");
    const body = nl === -1 ? "" : ch.slice(nl + 1);
    return parseTableRows(body);
  });
}

/** @returns {Record<string, { typ: string, det: string }>} */
function parse04(text) {
  const norm = text.replace(/\r\n/g, "\n");
  const map = {};
  for (const line of norm.split("\n")) {
    const t = line.trim();
    if (!t.startsWith("|")) continue;
    if (t.includes("占い名") && t.includes("出力方式")) continue;
    if (/^\|\s*[-:|]+\s*\|/.test(t)) continue;
    const cells = t
      .replace(/^\|/, "")
      .replace(/\|$/, "")
      .split("|")
      .map((c) => c.trim());
    if (cells.length < 2) continue;
    const [name, typ, det = ""] = cells;
    if (!name) continue;
    map[name] = { typ, det };
  }
  return map;
}

function colWidthUnits(str, maxW = 50) {
  if (!str) return 8;
  let w = 0;
  for (const ch of String(str)) {
    w += ch.codePointAt(0) < 128 ? 1 : 2;
  }
  return Math.min(Math.max(w * 0.55 + 2, 8), maxW);
}

const headers = [
  "No.",
  "占い名",
  "発祥地・文化圏",
  "カテゴリ",
  "必要な入力情報",
  "判定ロジック概要",
  "出力内容",
  "更新頻度",
  "実装難易度",
  "出力方式",
  "出力方式詳細",
  "実装メモ",
  "概要メモ",
];

const t1 = fs.readFileSync(FILE_01, "utf8");
const t2 = fs.readFileSync(FILE_02, "utf8");
const t4 = fs.readFileSync(FILE_04, "utf8");
const list01 = parse01(t1);
const list02 = parse02(t2);
const outType = parse04(t4);

if (list01.length !== list02.length) {
  console.error(`件数不一致: 01=${list01.length} 02=${list02.length}`);
  process.exit(1);
}

const wb = new ExcelJS.Workbook();
const ws = wb.addWorksheet("占い一覧", {
  views: [{ state: "frozen", ySplit: 1 }],
});

ws.addRow(headers);
const headerRow = ws.getRow(1);
headerRow.font = { name: "Meiryo UI", size: 11, bold: true };
headerRow.alignment = { vertical: "middle", horizontal: "center", wrapText: true };
headerRow.fill = {
  type: "pattern",
  pattern: "solid",
  fgColor: { argb: "FFCCCCCC" },
};
headerRow.height = 22;

const colMax = headers.map(() => 8);

for (let i = 0; i < headers.length; i++) {
  colMax[i] = Math.max(colMax[i], colWidthUnits(headers[i]));
}

for (let i = 0; i < list01.length; i++) {
  const d1 = list01[i];
  const d2 = list02[i];
  const r = i + 2;
  const name = d2["占い名"] || d1["占い名"] || "";
  const ot = outType[name] || { typ: "", det: "" };
  if (name && !outType[name]) {
    console.warn(`警告: 04_output_type.md に占い名がありません: ${JSON.stringify(name)}`);
  }
  const vals = [
    i + 1,
    name,
    d1["発祥地・文化圏"] || "",
    d1["出力カテゴリ"] || "",
    d2["必要な入力情報"] || "",
    d2["判定ロジック概要"] || "",
    d2["出力内容"] || "",
    d2["更新頻度"] || "",
    d2["実装難易度"] || "",
    ot.typ,
    ot.typ === "C" ? ot.det : "",
    d2["実装メモ"] || "",
    d1["概要メモ"] || "",
  ];
  ws.addRow(vals);
  const row = ws.getRow(r);
  row.font = { name: "Meiryo UI", size: 11 };
  row.alignment = { vertical: "top", wrapText: true };
  if (r % 2 === 0) {
    row.fill = {
      type: "pattern",
      pattern: "solid",
      fgColor: { argb: "FFF5F5F5" },
    };
  }
  vals.forEach((v, c) => {
    colMax[c] = Math.max(colMax[c], colWidthUnits(v));
  });
}

ws.columns = headers.map((_, i) => ({
  width: Math.min(colMax[i], 50),
}));

await wb.xlsx.writeFile(OUT);
console.log(`Wrote ${OUT} (${list01.length} data rows)`);
